7391 generate style properties on the fly based on wizard properties

juanignaciosl opened the pull request:
This closes #7391.

juanignaciosl commented:
@xavijam: this code generates a JS error when the form tries to initialize the style animated form (`this._animatedFormModel = new StyleAnimatedFormModel` is in the stack trace). Error:
```javascript
style-animated-properties-form-model.js:12 Uncaught TypeError: Cannot read property 'toString' of undefined
```

This is `style_properties` value:
```ruby
{"autogenerated"=>true,
 "type"=>"simple",
 "properties"=>
  {"blending"=>"none",
   "fill"=>{"color"=>{"fixed"=>"#FF6600", "opacity"=>0.9}},
   "labels"=>
    {"attribute"=>nil,
     "font"=>"DejaVu Sans Book",
     "offset"=>-10,
     "overlap"=>true,
     "placement"=>"simple",
     "fill"=>{"size"=>{"fixed"=>10}, "color"=>{"fixed"=>"#000", "opacity"=>1}},
     "halo"=>{"size"=>{"fixed"=>1}, "color"=>{"fixed"=>"#FFF", "opacity"=>1}},
     "enabled"=>true}}}
```

xavijam commented:
As we talked, we have to add default values for animated (and labels).

```rb
{"autogenerated"=>true,
 "type"=>"simple",
 "properties"=>
  {"blending"=>"none",
   "fill"=>{"color"=>{"fixed"=>"#FF6600", "opacity"=>0.9}},
   "labels"=>…,
   "animated"=>
    {"attribute"=>nil,
     "duration"=>30,
     "overlap"=>false,
     "steps"=>"256",
     "resolution"=>"2",
     "trails"=>"2",
     "enabled"=>false}}}
```

juanignaciosl commented:
@xavijam labels defaults (taken from the wiki, but with `enabled: false`):

```ruby
      labels: {
          enabled: false,
          attribute: null, // text-name
          font: 'DejaVu Sans Book', // text-face-name
          fill: {
            'size': {
              fixed: 10 // text-size
            },
            'color': {
              fixed: '#000', // text-fill
              opacity: 1
            }
          },
          halo: {
            'size': {
              fixed: 1 // text-halo-radius
            },
            'color': {
              fixed: '#111', // text-halo-fill
              opacity: 1
            }
          },
          offset: -10, // text-dy
          overlap: true, // text-allow-overlap
          placement: 'point' // text-placement-type
      }
```

xavijam commented:
@juanignaciosl, yep ^^^ 👍

juanignaciosl commented:
@javisantana Since there's no "cluster" mapping opening it with the new editor fails. This is the error:

```
POST http://juanignaciosl.localhost.lan:8181/api/v1/map?stat_tag=b2f1d664-20d1-11e6-80e3-080027880ca6&api_key=f5fe77ab2416fd66d4a6032d985b256fd411dd4e 400 (Bad Request)
map-base.js:80 Error creating the map instance on Windshaft: Postgis Plugin: ERROR:  column "points_count" does not exist
LINE 1: ..._webmercator",3.50861e-06), 7.01722e-06)) AS geom,"points_co...
                                                             ^

in executeQuery Full sql was: 'SELECT ST_AsBinary(ST_Simplify(ST_SnapToGrid("the_geom_webmercator",3.50861e-06), 7.01722e-06)) AS geom,"points_count","src" FROM (SELECT * FROM tac2016_stacked) as cdbq WHERE "the_geom_webmercator" && ST_MakeEnvelope(-20037508.3,20037508.25881302,-20037508.25881302,20037508.3,3857) LIMIT 65535'
```

Have you planned any strategy for cluster wizards?

javisantana commented:
No, cluster wizards should be:

1) marked as custom
2) move the query_wrapper to sql_wrap

juanignaciosl commented:
@gfiorav CR this, please.

@xavijam I haven't found major issues (other than missing _intensity_ wizard mapping). This is deployed to ded01, could you take a look to check that if I merge this it won't block you?

gfiorav commented on app/controllers/carto/api/layer_presenter.rb:
```diff
@@ -292,7 +302,445 @@ def whitelisted_infowindow(infowindow)
+      def migrate(options)
+        return nil unless @wizard_properties.present?
+
+        wpp = @wizard_properties['properties']
```
nit picking, but I don't like self explanatory variable names :/

juanignaciosl replied:
I know, but I need to use `wpp` and `spp` many times. They are `wizard_properties['properties']` and `style_properties['properties']`, respectively. I'm glad of replacing it for a better suggestion, as long as it's not long.

gfiorav commented on app/controllers/carto/api/layer_presenter.rb:
```diff
@@ -292,7 +302,445 @@ def whitelisted_infowindow(infowindow)
+        options['style_properties'] = {
+          'autogenerated' => true,
+          'type' => type,
+          'properties' => wizard_properties_properties_to_style_properties_properties(wpp)
```
You can de-rocket-operator-ize these like:
```ruby
{
  'patata': 'buena',
  'sin': 'cohetes'
}
```

up to you though!

(🚀 are so 2007)

juanignaciosl replied:
It's not equivalent, I think:

```ruby
irb(main):001:0> { 'patata' => 'buena' }
=> {"patata"=>"buena"}
irb(main):002:0> { 'patata': 'buena' }
=> {:patata=>"buena"}
irb(main):003:0> { 'patata' => 'buena' }['patata']
=> "buena"
irb(main):004:0> { 'patata': 'buena' }['patata']
=> nil
```

This time I _do_ want string keys, since it's a hash populated by the UI.

gfiorav commented on app/controllers/carto/api/layer_presenter.rb:
```diff
@@ -292,7 +302,445 @@ def whitelisted_infowindow(infowindow)
+
+      def set_if_present(hash, key, value)
+        # Dirty check because `false` is a valid `value`
+        hash[key] = value unless value.nil? || value == ''
```
You can use Rail's `.present?` to check for nullity/emptiness

juanignaciosl replied:
Ok, I know that Rails monkeypatching is a war I shouldn't have fought U_U

gfiorav commented on app/controllers/carto/api/layer_presenter.rb:
```diff
@@ -292,7 +302,445 @@ def whitelisted_infowindow(infowindow)
+      }.freeze
+
+      def wizard_properties_properties_to_style_properties_properties(wizard_properties_properties)
+        spp = {}
```
Same variable naming problem here

juanignaciosl replied:
My concern with that naming (I tried it before) is that it's miselading. Here's the structure of `layer.options['style_properties']` (same issue with `wizard_properties`):
```ruby
{
  kind: 'carto',
  style_properties: {
    type: 'choropleth',
    properties: {
        color: ...
    }
  }
}
```

As you see, my future me might thing that `style_ppties` might refer to `layer.options.style_properties` instead of ``layer.options.style_properties.properties`. That's why there're two `p` in `spp`: it's a contraction for `style_properties_properties`, which was the original name.

What do you think?

gfiorav commented:
Not much to add:

Nice and thorough tests. 📦 it!

juanignaciosl commented:
@gfiorav thanks!

@xavijam this is deployed in ded01 in case you want to take a look. I haven't found any major issue. I'll deploy it tomorrow morning.
